portal/serializers.py

```diff
@@ -0,0 +1,131 @@
+        validated_data["user"] = self.context["request"].user
+        # ensuring user cannot create an admin comment upon creation
+        validated_data["admin_comment"] = ""
+        validated_data["approved"] = False
+        return super().create(validated_data)
+
+    def update(self, instance, validated_data):
+        # if Poll is updated, then approve should be false
+        if not self.context["request"].user.is_superuser:
+            validated_data["approved"] = False
+        return super().update(instance, validated_data)
+
+
+class PollOptionSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = PollOption
+        fields = (
+            "id",
+            "poll",
+            "choice",
+        )
+
+    def update(self, instance, validated_data):
+        # if Poll Option is updated, then corresponding Poll approval should be false
+        instance.poll.approved = False
+        instance.poll.save()
+        return super().update(instance, validated_data)
+
+
+class RetrievePollSerializer(serializers.ModelSerializer):
+
+    target_populations = TargetPopulationSerializer(many=True)
+    options = PollOptionSerializer(source="polloption_set", many=True)
+
+    class Meta:
+        model = Poll
+        fields = (
+            "id",
+            "source",
+            "question",
+            "image_url",
+            "created_date",
+            "expire_date",
+            "multiselect",
+            "user_comment",
+            "options",
+            "target_populations",
+        )
+
+
+class PollVoteSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = PollVote
+        fields = ("id", "poll_options", "created_date")
+        read_only_fields = ("created_date",)
+
+    def create(self, validated_data):
+
+        options = validated_data["poll_options"]
+        poll = options[0].poll
+        # checks if user has already voted
+        if PollVote.objects.filter(user=self.context["request"].user, poll=poll).exists():
+            print(PollVote.objects.filter(user=self.context["request"].user))
+            raise serializers.ValidationError(
+                detail={"detail": "You have already voted for this Poll"}
+            )
+        # check if user can multiselect or not
+        if len(options) > 1 and not poll.multiselect:
+            raise serializers.ValidationError(
+                detail={"detail": "You cannot select multiple choices for this Poll"}
+            )
+        # check if poll options are all from same Poll
+        for option in options:
+            if option.poll != poll:
+                raise serializers.ValidationError(
+                    detail={"detail": "Voting options are from different Polls"}
+                )
+        # checks if user belongs to target population
+        if not any(
+            i in list(poll.target_populations.all().values_list("id", flat=True))
+            for i in get_user_populations(self.context["request"].user)
```

Comment:
I'm pretty sure this is an inefficient way of doing this check. Looks like you iterate over all poll target populations for each user population, right?

Reply:
Bump on this. Would something like `poll.target_populations.filter(id__in=get_user_populations(self.context["request"].user).count == 0` not work?